cpp/oneapi/dal/algo/svm/backend/gpu/smo_solver.hpp

```diff
@@ -0,0 +1,280 @@
+/*******************************************************************************
+* Copyright 2021 Intel Corporation
+*
+* Licensed under the Apache License, Version 2.0 (the "License");
+* you may not use this file except in compliance with the License.
+* You may obtain a copy of the License at
+*
+*     http://www.apache.org/licenses/LICENSE-2.0
+*
+* Unless required by applicable law or agreed to in writing, software
+* distributed under the License is distributed on an "AS IS" BASIS,
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*******************************************************************************/
+
+#pragma once
+
+#include "oneapi/dal/backend/primitives/ndarray.hpp"
+#include "oneapi/dal/backend/primitives/common.hpp"
+#include "oneapi/dal/algo/svm/backend/gpu/misc.hpp"
+
+namespace oneapi::dal::svm::backend {
+
+namespace pr = dal::backend::primitives;
+
+using sycl::ONEAPI::reduce;
+using sycl::ONEAPI::maximum;
+using sycl::ONEAPI::minimum;
+
+template <typename Data>
+using local_accessor_rw_t =
+    sycl::accessor<Data, 1, sycl::access::mode::read_write, sycl::access::target::local>;
+
+template <typename Float>
+struct enumerate_value {
+    std::uint32_t index;
+    Float value;
+};
+
+template <typename Float>
+struct local_variables {
+    Float delta_b_i;
+    Float delta_b_j;
+    Float local_diff;
+    Float local_eps;
+    enumerate_value<Float> max_val_ind;
+};
+
+template <typename Float>
+inline void reduce_arg_max(sycl::nd_item<1> item,
```

Comment:
[optional] This functionality can be useful not only here. Maybe think about where you could put in a common place?

Reply:
Where else is it used?

Reply:
For common place - maybe primitives... to use in other algorithms